oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+
+If two modules require incompatible versions of a shared dependency, the second one loaded can fall back to loading a version it's compatible with from its own build. (see the link in step #4 below for details.)
+
+In terms of Open edX MFEs, this means:
+
+1. MFEs can continue to be built independently.
+2. The Webpack build will include a manifest of which sub-modules the MFE provides at runtime.
+3. ``@openedx/frontend-build`` will specify - through its Webpack configs - a common set of shared dependencies to be used at runtime by all MFEs.
+4. Webpack will intelligently resolve those dependencies at runtime, `taking into account each module's specific version requirements <https://www.angulararchitects.io/en/blog/getting-out-of-version-mismatch-hell-with-module-federation>`_.
+5. MFEs can dynamically load modules from other MFEs at runtime with Webpack handling hooking them up to the right dependencies.
+
+Because we already use Webpack, the work to add the ModuleFederationPlugin to our configurations is small and uninvasive (see proof of concept in the `Reference Implementation`_ section below).
+
+Maintaining Dependency Consistency
+----------------------------------
+
+This proposal fundamentally changes how we work with MFE dependencies, and will require us to adopt a more rigorous approach to ensuring dependency consistency and compatiblity. That consistency doesn't come for free just by adopting shared dependencies at runtime: independent codebases will need to be kept in sync.
+
+We expect that this may need to take a number of possible forms.
+
+Process
+^^^^^^^
+
+We need to ensure maintainers and developers know what dependency versions to use, and when they need to upgrade to stay consistent. Open edX release documentation should include documentation of which frontend dependency versions are compatible with the release, likely pinned to a major version (i.e., React 17.x, Paragon 22.x, etc.)
+
+We also need a process to migrate core repositories through breaking changes in third-party dependencies. Ideally follow the `Upgrade Project Runbook <https://openedx.atlassian.net/wiki/spaces/AC/pages/3660316693/Upgrade+Project+Runbook>`_.
+
+Best Practices
+^^^^^^^^^^^^^^
+
+We need to ensure we minimize breaking changes in our own libraries (such as Paragon, the header, footer, frontend-platform, frontend-build, etc.) We suggest accomplishing this by:
+
+- Creating new versions of components with breaking changes (``ButtonV2``, ``webpack.dev.config.v2.js``) rather than modifying existing ones.
+- Leveraging the DEPR process for communication and removing old component versions
+- Aligning that removal and the subsequent breaking changes with Open edX releases, and documenting it in their release notes.
+
+Further, we could reduce the overhead of dependency maintenance and ensure MFEs stay up to date by pinning dependencies to major versions using ``^`` on versions in our package.json files.
```

Comment:
Is this a recommendation? Are we following this? Or is this thinking out loud?

Reply:
*I* have been doing this when updating dependencies but I still see quite a bit of exact dependency pins making it in. Might be something we should make an ADR under https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0067-bp-tools-and-technology.html

Reply:
Yeah, this sounds to me like a good practice that we should just codify and follow.